mkosi/run.py

```diff
@@ -359,101 +403,91 @@ def bwrap_cmd(
-    network: bool = False,
-    stdout: _FILE = None,
-    env: Mapping[str, PathString] = {},
-) -> CompletedProcess:
+def chroot_cmd(root: Path, *, options: Sequence[PathString] = (), network: bool = False) -> Sequence[PathString]:
     cmdline: list[PathString] = [
         "bwrap",
         "--unshare-ipc",
         "--unshare-pid",
         "--unshare-cgroup",
-        "--bind", root, "/",
-        "--tmpfs", "/run",
-        "--tmpfs", "/tmp",
-        "--dev", "/dev",
-        "--proc", "/proc",
-        "--ro-bind", "/sys", "/sys",
+        "--dev-bind", root, "/",
         "--die-with-parent",
-        *bwrap_params,
+        "--setenv", "container", "mkosi",
+        "--setenv", "SYSTEMD_OFFLINE", str(int(network)),
+        "--setenv", "HOME", "/",
+        "--setenv", "PATH", "/usr/bin:/usr/sbin",
+        *options,
     ]
 
-    resolve = root.joinpath("etc/resolv.conf")
-
-    tmp = Path(tempfile.NamedTemporaryFile(delete=False).name)
-    tmp.unlink()
-
     if network:
-        # Bubblewrap does not mount over symlinks and /etc/resolv.conf might be a symlink. Deal with this by
-        # temporarily moving the file somewhere else.
-        if resolve.is_symlink():
-            shutil.move(resolve, tmp)
+        resolve = Path("etc/resolv.conf")
+        if (root / resolve).is_symlink():
+            # bubblewrap cannot mount over symlinks so handle that manually here.
+            resolve = (root / resolve).readlink()
 
-        # If we're using the host network namespace, use the same resolver
-        cmdline += ["--ro-bind", "/etc/resolv.conf", "/etc/resolv.conf"]
+        # If we're using the host network namespace, use the same resolver.
+        cmdline += ["--ro-bind", "/etc/resolv.conf", Path("/") / resolve]
     else:
         cmdline += ["--unshare-net"]
 
-    env = dict(
-        container="mkosi",
-        SYSTEMD_OFFLINE=str(int(network)),
-        HOME="/",
-        PATH="/usr/bin:/usr/sbin",
-    ) | env
+    return cmdline
 
-    with tempfile.TemporaryDirectory(dir="/var/tmp", prefix="mkosi-var-tmp") as var_tmp:
-        cmdline += [
-            "--bind", var_tmp, "/var/tmp",
-            "sh", "-c", "chmod 1777 /tmp /var/tmp /dev/shm && exec $0 \"$@\" || exit $?"
-        ]
 
-        try:
-            return run([*cmdline, *cmd], text=True, stdout=stdout, env=env, log=False)
-        except subprocess.CalledProcessError as e:
-            logging.error(f"\"{' '.join(str(s) for s in cmd)}\" returned non-zero exit code {e.returncode}.")
-            if ARG_DEBUG_SHELL.get():
-                run([*cmdline, "sh"], stdin=sys.stdin, check=False, env=env, log=False)
-            raise e
-        finally:
-            if tmp.is_symlink():
-                resolve.unlink(missing_ok=True)
-                shutil.move(tmp, resolve)
+def which(program: str, tools: Optional[Path]) -> Optional[Path]:
+    result = bwrap(["sh", "-c", f"command -v {program}"], tools=tools, check=False, stdout=subprocess.PIPE)
```

Comment:
Instead of shelling out for each which, why not call `shutil.which` with the `path` argument? This would be a bit more lightweight than starting the whole bwrap machinery and a shell every time.

Reply:
Because we want our implementation of `which()` to run in the same environment the command will be executed in. I think it's valuable to match the environment as much as possible here.

We could use `which` instead of the shell but that's another tool that needs to be installed in the tools tree whether the shell is already needed for other things.

Reply:
I generally agree that we want to match the execution environment as closely as possible, but the only thing `shutil.which` does is go through the files in a list of directories and return the first matching the given name that is both readable and executable. I assume `command` does little else, so I'm not sure there is added value to set up another execution environment when we can make the same decision for the same paths without.